### PR TITLE
chore: Refactor Source to Avoid Implementing Forwarder Interface 

### LINKED ISSUE #1663: Refactor Source Implementations to Avoid Implementing Forwarder Interface
Since we have a separate forwarder for the source, our sources don't need to implement the forwarder interface.

Similar to #1662

### PR DESCRIPTION
**Pull request by yhl25**
fixes #1663

### CODE REVIEW

File: pkg/sources/forward/data_forward.go [hunk 1]
```diff
@@ -270,6 +270,7 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
 		messageToStep[toVertex] = make([][]isb.Message, len(df.toBuffers[toVertex]))
 	}
 
+	// FIXME: when the transformer is not defined, we should avoid doing this.
```
Comment (KeranYang) [inline pkg/sources/forward/data_forward.go:273]: yes. @ayildirim21 is working on this.

File: pkg/sources/source.go [hunk 2]
```diff
@@ -200,69 +207,90 @@ func (sp *SourceProcessor) Start(ctx context.Context) error {
...
+
 	if sp.VertexInstance.Vertex.HasUDTransformer() {
 		// Wait for server info to be ready
 		serverInfo, err := sdkserverinfo.SDKServerInfo(sdkserverinfo.WithServerInfoFilePath(sdkclient.SourceTransformerServerInfoFile))
 		if err != nil {
 			return err
 		}
 
-		sdkClient, err = sourcetransformer.New(serverInfo, sdkclient.WithMaxMessageSize(maxMessageSize))
+		sdkClient, err := sourcetransformer.New(serverInfo, sdkclient.WithMaxMessageSize(maxMessageSize))
```
Comment (KeranYang) [inline pkg/sources/source.go:232]:
```suggestion
		srcTranformerClient, err := sourcetransformer.New(serverInfo, sdkclient.WithMaxMessageSize(maxMessageSize))
```

File: pkg/sources/source.go [hunk 3]
```diff
@@ -200,69 +207,90 @@ func (sp *SourceProcessor) Start(ctx context.Context) error {
...
 		if err != nil {
-			return fmt.Errorf("failed to create gRPC client, %w", err)
+			return fmt.Errorf("failed to create transformer gRPC client, %w", err)
 		}
 
-		transformerGRPCClient := transformer.NewGRPCBasedTransformer(sdkClient)
+		transformerClient = transformer.NewGRPCBasedTransformer(sdkClient)
```
Comment (KeranYang) [inline pkg/sources/source.go:237]:
```suggestion
		srcTransformerGRPCClient = transformer.NewGRPCBasedTransformer(sdkClient)
```

nit: this is to be consistent with srcClient/udsGRPCClient above for udsource.

**Review by KeranYang — COMMENTED** [review]
LGTM. I think we can merge once UT failures are fixed.

File: pkg/sources/kafka/reader.go [hunk 4]
```diff
@@ -28,107 +28,109 @@ import (
...
+			return nil, err
+		}
+	}
 
-// AckIt acking is taken care by the consumer group
-func (k *kafkaOffset) AckIt() error {
-	// NOOP
-	return nil
-}
+	sarama.NewConfig()
```
Comment (whynowy) [inline pkg/sources/kafka/reader.go:76]: Is this line in-use? How was it introduced?

File: pkg/sources/sourcer/sourcer.go [hunk 5]
```diff
@@ -40,12 +39,9 @@ type SourceReader interface {
...
 	Partitions(ctx context.Context) []int32
+	// Pending returns the pending messages number.
+	Pending(context.Context) (int64, error)
 }
 
-// Sourcer interface provides an isb.BufferReader abstraction over the underlying data source.
-// This is intended to be consumed by a connector like isb.forward
-type Sourcer interface {
-	SourceReader
-	forwarder.StarterStopper
```
Comment (whynowy) [inline pkg/sources/sourcer/sourcer.go:49]: In this case, how to implement those logics for `preStop`, `preStart` or `postStart`? Like the `ready` flag used in HTTP source?

I understand for `preStart` or `postStart`, we could leverage the `New` function to sort of achieve (e.g. you start the kafka consumer Goroutine) that, however that is not an elegant implementation, and it relies on the `ctx` cancel to exit, and there's no way to do a clean shutdown.

Comment (whynowy) [reply pkg/sources/sourcer/sourcer.go:49]: Synced up offline, we could go ahead.

**Comment by vigith** [comment]
Though this PR removes some redundant code required for implementing a built-in source, this PR does not solve the problem of orchestrating a clean shutdown. An ideal source reader will always read ahead and be ready to hand over data, but when shutting down, it will ensure the read-ahead data is emptied.

Perhaps we need to rethink how to orchestrate a clean shutdown for a source (also for a sink), and this PR could remove the redundant code **without any regression**.

Using `ctx.Done` to do shutdown is like having a bull run through a shop. It will randomly pull down things without any proper order, and we will have to heavily rely on custom edge case handling.

**Comment by vigith** [comment]
@whynowy how should we proceed? Should we merge this in and then make it cleaner in subsequent PRs?

**Review by whynowy — APPROVED** [review]

**Pull request merged**